src/server/pfs/server/driver.go

```diff
@@ -1790,17 +1780,54 @@ func (d *driver) listCommitF(pachClient *client.APIClient, repo *pfs.Repo, to *p
 		return fmt.Errorf("cannot use `from` commit without `to` commit")
 	} else if from == nil && to == nil {
 		// if neither from and to is given, we list all commits in
-		// the repo, sorted by revision timestamp
-		if err := commits.List(ci, col.DefaultOptions, func(commitID string) error {
-			if number <= 0 {
-				return errutil.ErrBreak
+		// the repo, sorted by revision timestamp (or reversed if so requested.)
+		opts := *col.DefaultOptions // Note we dereference here so as to make a copy
+		if reverse {
+			opts.Order = etcd.SortAscend
+		}
+		// we hold onto a revisions worth of cis so that we can sort them by provenance
+		var cis []*pfs.CommitInfo
+		// sendCis sorts cis and passes them to f
+		sendCis := func() error {
+			// Sort in reverse provenance order, i.e. commits come before their provenance
+			sort.Slice(cis, func(i, j int) bool { return len(cis[i].Provenance) > len(cis[j].Provenance) })
+			for i, ci := range cis {
+				if number <= 0 {
+					return errutil.ErrBreak
+				}
+				number--
+
+				if reverse {
+					ci = cis[len(cis)-1-i]
```

Comment:
It's hard for me to tell, but should this be a swap? In other words,
```
last := len(cis)-1
cis[i], cis[last-i] = cis[last-i], cis[i]
```
Right now it looks like the comments at the beginning are overwritten?

Reply:
Yup, I'm just doing reverse iteration here.